src/providers/pkcs11_provider/utils.rs

```diff
@@ -1,14 +1,275 @@
 // Copyright 2020 Contributors to the Parsec project.
 // SPDX-License-Identifier: Apache-2.0
-
 use super::Pkcs11Provider;
 use log::error;
 use log::{info, trace, warn};
+use parsec_interface::operations::psa_algorithm::*;
+use parsec_interface::operations::psa_key_attributes::*;
 use parsec_interface::requests::ResponseStatus;
 use parsec_interface::requests::Result;
+use picky_asn1_x509::{AlgorithmIdentifier, DigestInfo, SHAVariant};
 use pkcs11::errors::Error;
 use pkcs11::types::*;
 use pkcs11::types::{CKF_RW_SESSION, CKF_SERIAL_SESSION, CKU_USER};
+use std::convert::{TryFrom, TryInto};
+use std::pin::Pin;
+
+// Public exponent value for all RSA keys.
+const PUBLIC_EXPONENT: [u8; 3] = [0x01, 0x00, 0x01];
+
+/// Abstraction over CK_MECHANISM_TYPE
+#[derive(Debug, Copy, Clone)]
+pub enum CkMechanismType {
+    CkmRsaPkcs,
+    CkmRsaPkcsPss,
+    CkmRsaPkcsOaep,
+    CkmSha1,
+    CkmSha256,
+    CkmSha384,
+    CkmSha512,
+}
+
+impl From<CkMechanismType> for CK_MECHANISM_TYPE {
+    fn from(mech_type: CkMechanismType) -> Self {
+        match mech_type {
+            CkMechanismType::CkmRsaPkcs => CKM_RSA_PKCS,
+            CkMechanismType::CkmRsaPkcsPss => CKM_RSA_PKCS_PSS,
+            CkMechanismType::CkmRsaPkcsOaep => CKM_RSA_PKCS_OAEP,
+            CkMechanismType::CkmSha1 => CKM_SHA_1,
+            CkMechanismType::CkmSha256 => CKM_SHA256,
+            CkMechanismType::CkmSha384 => CKM_SHA384,
+            CkMechanismType::CkmSha512 => CKM_SHA512,
+        }
+    }
+}
+
+pub fn mech_type_to_allowed_mech_attribute(mech_type: &mut CK_MECHANISM_TYPE) -> CK_ATTRIBUTE {
```

Comment:
Can the `CK_MECHANISM` referenced here move after this function call? I would say not because of the mutable referenced that is passed.